tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -1616,6 +1591,198 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
+				err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+					tc.Spec.PD.Replicas = 3
+					return nil
+				})
+				framework.ExpectNoError(err, "failed to scale in PD for TidbCluster %s/%s", ns, tc.Name)
+
+				ginkgo.By("Wait for PD to be in ScalePhase")
+				utiltc.MustWaitForPDPhase(cli, tc, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+
+				ginkgo.By("Upgrade PD version concurrently")
+				err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+					tc.Spec.PD.Version = pointer.StringPtr(utilimage.TiDBV4)
+					return nil
+				})
+				framework.ExpectNoError(err, "failed to scale in PD for TidbCluster %s/%s", ns, tc.Name)
+
+				ginkgo.By("Wait for tc ready")
+				err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 10*time.Second)
+				framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s ready after scale in pd", ns, tc.Name)
+
+				ginkgo.By("Check PD Pods")
+				listOptions := metav1.ListOptions{
+					LabelSelector: labels.SelectorFromSet(label.New().Instance(tc.Name).PD().Labels()).String(),
+				}
+				pods, err := c.CoreV1().Pods(ns).List(listOptions)
+				framework.ExpectNoError(err, "failed to list PD Pods with options: %+v", listOptions)
+				framework.ExpectEqual(len(pods.Items), 3, "there should be 3 PD Pods")
+				for _, pod := range pods.Items {
+					framework.ExpectEqual(fmt.Sprintf("pingcap/pd:%s", utilimage.TiDBV4), pod.Spec.Containers[0].Image, "PD Pod has wrong image")
+				}
+			})
+		})
+
+		ginkgo.It("PD to 0 is forbidden while other components are running", func() {
+			ginkgo.By("Deploy initial tc")
+			tc := fixture.GetTidbCluster(ns, "scale-pd-to-0", utilimage.TiDBV4)
+			tc.Spec.PD.Replicas = 1
+			utiltc.MustCreateTCWithComponentsReady(genericCli, oa, tc, 5*time.Minute, 10*time.Second)
+
+			ginkgo.By("Scale in PD to 0 replicas")
+			err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+				tc.Spec.PD.Replicas = 0
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to scale in PD for TidbCluster %s/%s", ns, tc.Name)
+
+			ginkgo.By("Wait for PD to be in ScalePhase")
+			utiltc.MustWaitForPDPhase(cli, tc, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+
+			ginkgo.By("Check for FailedScaleIn event")
+			// LAST SEEN   TYPE      REASON          OBJECT              MESSAGE
+			// 25s         Warning   FailedScaleIn   tidbcluster/basic   The PD is in use by TidbCluster [pingcap/basic], can't scale in PD, podname basic-pd-0
+			err = wait.Poll(5*time.Second, 1*time.Minute, func() (done bool, err error) {
+				options := metav1.ListOptions{
+					FieldSelector: fmt.Sprintf("involvedObject.name=%s,involvedObject.kind=TidbCluster", tc.Name),
+				}
+				event, err := c.CoreV1().Events(ns).List(options)
+				if err != nil {
+					log.Logf("failed to list events with options: +%v", options)
+					return false, nil
+				}
+				for _, event := range event.Items {
+					log.Logf("found event: %+v", event)
+					if event.Reason == "FailedScaleIn" {
```

Comment:
How can we make sure that this event is for **PD** in **this TC**?

Reply:
added message matching with "PD" for now